File: backend/controllers/order_controller.go
package controllers
import (
	"net/http"
	"cronos-backend/middleware"
	"cronos-backend/services"
	"cronos-backend/utils"
	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)
type OrderController struct {
	OrderService *services.OrderService
}
func NewOrderController(service *services.OrderService) *OrderController {
	return &OrderController{OrderService: service}
}
func (ctrl *OrderController) CreateOrder(c *gin.Context) {
	userID := middleware.GetUserID(c)
	var req services.CreateOrderRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.BadRequest(c, "Invalid request body", err.Error())
		return
	}
	order, err := ctrl.OrderService.CreateOrder(userID, req)
	if err != nil {
		utils.BadRequest(c, err.Error(), nil)
		return
	}
	utils.SuccessResponse(c, http.StatusCreated, "Order created", order)
}
func (ctrl *OrderController) GetMyOrders(c *gin.Context) {
	userID := middleware.GetUserID(c)
	orders, meta, err := ctrl.OrderService.GetMyOrders(userID, c.Query("page"), c.Query("limit"))
	if err != nil {
		utils.InternalServerError(c, err.Error())
		return
	}
	utils.SuccessResponseWithMeta(c, http.StatusOK, "Orders retrieved", orders, meta)
}
func (ctrl *OrderController) GetAllOrders(c *gin.Context) {
	orders, meta, err := ctrl.OrderService.GetAllOrders(c.Query("page"), c.Query("limit"), c.Query("status"))
	if err != nil {
		utils.InternalServerError(c, err.Error())
		return
	}
	utils.SuccessResponseWithMeta(c, http.StatusOK, "Orders retrieved", orders, meta)
}
func (ctrl *OrderController) GetOrder(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		utils.BadRequest(c, "Invalid order ID", nil)
		return
	}
	order, err := ctrl.OrderService.GetOrder(id)
	if err != nil {
		utils.NotFound(c, "Order not found")
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "Order retrieved", order)
}
func (ctrl *OrderController) CancelOrder(c *gin.Context) {
	userID := middleware.GetUserID(c)
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		utils.BadRequest(c, "Invalid order ID", nil)
		return
	}
	if err := ctrl.OrderService.CancelOrder(userID, id); err != nil {
		utils.BadRequest(c, err.Error(), nil)
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "Order cancelled", nil)
}
func (ctrl *OrderController) GetSellerOrders(c *gin.Context) {
	userID := middleware.GetUserID(c)
	orders, err := ctrl.OrderService.GetSellerOrders(userID)
	if err != nil {
		utils.InternalServerError(c, err.Error())
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "Seller orders retrieved", orders)
}

File: backend-spring/src/main/java/com/cronos/controller/OrderController.java
package com.cronos.controller;

import com.cronos.dto.request.CreateOrderRequest;
import com.cronos.dto.response.ApiResponse;
import com.cronos.service.OrderService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;
import java.util.UUID;

@RestController
@RequestMapping("/api/orders")
@RequiredArgsConstructor
public class OrderController {
    private final OrderService orderService;

    @PostMapping
    public ResponseEntity<?> create(Authentication auth, @Valid @RequestBody CreateOrderRequest req) {
        try {
            return ResponseEntity.ok(ApiResponse.success("Order created", orderService.createOrder((UUID) auth.getPrincipal(), req)));
        } catch (RuntimeException e) { return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage())); }
    }

    @GetMapping
    public ResponseEntity<?> getMyOrders(Authentication auth, @RequestParam(defaultValue = "1") String page, @RequestParam(defaultValue = "10") String limit) {
        var result = orderService.getMyOrders((UUID) auth.getPrincipal(), page, limit);
        return ResponseEntity.ok(ApiResponse.success("Orders retrieved", result.get("orders"), result.get("meta")));
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getOrder(@PathVariable UUID id) {
        return ResponseEntity.ok(ApiResponse.success("Order retrieved", orderService.getOrder(id)));
    }

    @PutMapping("/{id}/cancel")
    public ResponseEntity<?> cancel(Authentication auth, @PathVariable UUID id) {
        try {
            orderService.cancelOrder((UUID) auth.getPrincipal(), id);
            return ResponseEntity.ok(ApiResponse.success("Order cancelled", null));
        } catch (RuntimeException e) { return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage())); }
    }
}
